crates/sui-indexer-alt-graphql/src/api/types/available_range.rs

```diff
@@ -0,0 +1,382 @@
+///
+/// Both `type_` and `field` are required. The `filter` is optional and provides retention information for filtered queries.
+#[derive(InputObject, Debug, Clone, Eq, PartialEq)]
+pub(crate) struct RetentionKey {
+    /// The GraphQL type to check retention for
+    pub(crate) type_: String,
+
+    /// The specific field within the type to check retention for
+    pub(crate) field: String,
+
+    /// Optional filter to check retention for filtered queries
+    pub(crate) filters: Option<Vec<String>>,
+}
+
+#[derive(Clone)]
+pub struct AvailableRange {
+    pub scope: Scope,
+    pub first: u64,
+}
+
+/// Checkpoint range for which data is available.
+#[Object]
+impl AvailableRange {
+    /// Inclusive lower checkpoint for which data is available.
+    async fn first(&self) -> Result<Option<Checkpoint>, RpcError> {
+        Ok(Checkpoint::with_sequence_number(
+            self.scope.clone(),
+            Some(self.first),
+        ))
+    }
+
+    /// Inclusive upper checkpoint for which data is available.
+    async fn last(&self) -> Result<Option<Checkpoint>, RpcError> {
+        Ok(Checkpoint::with_sequence_number(self.scope.clone(), None))
+    }
+}
+
+impl AvailableRange {
+    /// Get retention information for a specific query type and field
+    pub(crate) fn new(
+        ctx: &Context<'_>,
+        scope: &Scope,
+        retention_key: RetentionKey,
+    ) -> Result<Self, RpcError> {
+        let watermarks: &Arc<Watermarks> = ctx.data()?;
+        let pipelines = pipelines(
+            &retention_key.type_,
+            &retention_key.field,
+            retention_key.filters,
+        );
+
+        let lo_checkpoint =
+            pipelines
+                .iter()
+                .try_fold(0, |acc: u64, pipeline| -> Result<u64, RpcError> {
+                    let watermark = watermarks.pipeline_lo_watermark(pipeline)?;
+                    let checkpoint = watermark.checkpoint();
+                    Ok(acc.max(checkpoint))
+                })?;
+
+        Ok(Self {
+            scope: scope.clone(),
+            first: lo_checkpoint,
+        })
+    }
+}
+
+/// Maps GraphQL query components to watermark pipeline names.
+///
+/// Determines which watermark pipelines are relevant for a given GraphQL query.
+/// The pipeline names are used to query watermark data to determine the
+/// checkpoint sequence range (available range) for which data is available.
+///
+fn pipelines(type_: &str, field: &str, filters: Option<Vec<String>>) -> Vec<&'static str> {
+    let mut filters = filters.unwrap_or_default();
+    match (type_, field) {
+        // Address fields
+        ("Address", "address") => vec![],
+        ("Address", "asObject") => vec!["obj_versions"],
+        ("Address", "balance") => pipelines("IAddressable", "balance", None),
+        ("Address", "balances") => pipelines("IAddressable", "balances", None),
+        ("Address", "defaultSuiNsName") => pipelines("IAddressable", "defaultSuiNsName", None),
```

Comment:
(and everywhere else)

```suggestion
        ("Address", "defaultSuinsName") => pipelines("IAddressable", "defaultSuinsName", None),
```